contributors/design-proposals/apps/configmap-secret-gc.md

```diff
@@ -0,0 +1,271 @@
+        previously referenced but no longer referenced ConfigMaps/Secrets. 
+
+
+## Alternatives
+
+Other alternatives considered are listed below. 
+
+
+### Embed ConfigMaps/Secrets in Controller Spec
+
+Add `.configMapTemplate` and `.secretTemplate` fields to a controller's `.spec`. 
+
+**Cons:**
+
+*   Adds complexity to the API, tooling dealing with ConfigMaps/Secrets, and the
+    implementation of controllers.
+*   Controller spec becomes fat.
+
+
+### Specify ConfigMaps/Secrets Garbage Collection in Controller Spec
+
+Add a `bool` field in controller `.spec` to specify whether garbage collection
+of its ConfigMaps/Secrets is enabled. For more granularity, add another field
+allowing users to specify a list of ConfigMaps/Secrets that should not be
+garbage collected even if the aforementioned field is enabled. 
+
+**Cons:**
+
+*   User intent conflict. For example, user A wants a ConfigMap to be garbage
+    collected with a Deployment, and therefore the ConfigMap will be deleted
+    after the Deployment is gone. User B cannot stop the ConfigMap from being
+    deleted even if B disables garbage collection of that ConfigMap in B's
+    controller spec. 
+*   To enable or disable garbage collection of a ConfigMap/Secret, users need to
+    update spec of all controllers that reference it. 
+
+
+### Specify ConfigMaps/Secrets garbage collection in PodSpec
+
+Add a `bool` field everywhere in `PodSpec` that ConfigMaps/Secrets are
+referenced (e.g. `.containers.envFrom.configMapRef`) to specify whether garbage
+collection of the ConfigMap/Secret is enabled. 
+
+**Cons:**
+
+*   Verbosity. if a ConfigMap is referenced in multiple places in a pod template,
+    users need to specify it everywhere. 
+*   User intent conflict.
+*   To enable or disable garbage collection of a ConfigMap/Secret, users need to
+    update pod template of all controllers that reference it. For controllers
+    with history, this means triggering rollouts.
+
+
+### Automatic ConfigMap/Secret References Updates
+
+Every time a ConfigMap/Secret is updated, the controller automatically creates a
+new ConfigMap/Secret (the new name can be generated by appending a hash), and
+updates references in pod templates in controller history objects. Every
+ConfigMap/Secret update triggers a rollout.
+
+For example, make Deployment controller create an ConfigMap "abc-<hash>" from
+ConfigMap "abc" referenced in its pod template, and then reference "abc-<hash>"
+in ReplicaSet pod template. Or something like
+[#31701](https://github.com/kubernetes/kubernetes/pull/31701).
+
+This is not strictly about garbage collection of ConfigMaps/Secrets, but is
+relevant to the garbage collection design. 
+
+**Pros:**
+
+*   ConfigMap/Secret references are updated automatically.
+
+**Cons:**
+
+*   It becomes tricky for controllers (e.g. Deployments) to compare its pod
+    template against its history (e.g. ReplicaSets) to find current history
```

Comment:
Have something mutate the deployment - which avoids this problem.  Alternatively, make the deployment tolerate refinement of its ReplicaSets by admission and initializers (this is inevitable - image resolution is one concrete example we'll get asked to do).

Reply:
If something outside mutates the Deployment, will we need a way to prevent the Deployment controller from seeing the original CM ref before it's resolved to point to a copy?

Reply:
I think the trigger controller (which is what I would call the thing acting on the resource) is the one who is resolving refs.  That means the none of the existing controllers change.

I partially am advocating for this approach because I have direct familiarity with it from several years of working on the openshift image triggering mechanism.  The mechanism is:

1. a resource exposes a "latest state value" (an image stream tag, which points to a concrete image by digest)
2. a controller that reconciles the latest state value with the status of a deployment

https://github.com/openshift/origin/blob/master/pkg/image/controller/trigger/image_trigger_controller.go
https://github.com/openshift/origin/blob/master/pkg/image/trigger/annotations/annotations.go
https://github.com/openshift/origin/blob/master/pkg/oc/cli/cmd/set/triggers.go#L809

The user sets an annotation on any object (deployment, stateful set, daemon set, replica set) pointing to an image stream tag (roughly `foo/bar:latest` which resolves to `sha256:abc`).  The controller is level driven and ensures that all resources with that annotation have the correct resolved value in their config.  The mechanism can be paused by updating the annotation.  Otherwise, any watched resource can be mutated.

The key advantages are that this can be extended to any object (it could be used for an etcd operator, a new workload controller, etc) just by adding the annotation.  The downside is that the object is under automatic control.  However, this is similar to the autoscaler - a controller is programming down onto the object and the object is unaware of its programming.

A similar model for secret and config maps would be:

1. support an annotation on the deployment that indicates it wishes to receive the latest copy of a configmap, and give a volume name (or jsonpath, or other target) to replace
2. the controller watches the configmap and ensures
    1. a versioned configmap exists for every targeted configmap
    2. every resource referencing that versioned configmap is pointing to the latest copy of the configmap

Rollback is accomplished by reverting the old configmap, pausing the trigger on a per resource basis, or removing the annotation.  Same advantage exists here - no changes to existing controllers is required, all logic is centralized, and all extension / CRD objects can benefit from this.

The per controller approach has merits, but I think it's biggest weakness is that it requires a large amount of duplicate code that can be better centralized.  Annotations are annoying - but if necessary we could introduce future changes that make this a formal field on an object (it's unlikely a common schema exists), or we could place it on a config map.

Reply:
The controller model for this is roughly what is described by @janetkuo down below, just automated.  I believe it's a natural next step to automate (we had been planning on doing just that for other reasons), which makes me lean towards GC mechanisms that align with that model.

Reply:
How does GC work in this case? The solution @janetkuo is proposing ensures that the owner references of the config objects point to the objects that implement history for a controller. As long as the config is within the revision history limit you can be assured that it still exists and that you can roll back to it. Also, how do you ensure that the config injection controller updates the configuration of the ReplicaSet prior to the ReplicaSet controller materializing Pods that point to a nil configuration.

Reply:
I agree with @kow3ns. Whether the references are updated by a controller or by a non-continuously running client, the GC implementation is orthogonal and the same.

Reply:
Updating the image field to point at the digest in the deployment pod template is analogous to updating references to ConfigMaps and Secrets.